t/loss.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2017 Fastly, Kazuho Oku
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include "quicly/loss.h"
+#include "quicly/defaults.h"
+#include "test.h"
+
+#define ASSERT(expr)                                                                                                               \
+    do {                                                                                                                           \
+        int ret = !!(expr);                                                                                                        \
+        assert(ret);                                                                                                               \
+    } while (0)
+
+static int64_t now;
+static uint64_t num_packets_lost = 0;
+
+static void on_loss_detected(quicly_loss_t *loss, const quicly_sent_packet_t *lost_packet, int is_time_threshold)
+{
+    ++num_packets_lost;
+}
+
+static void acked(quicly_loss_t *loss, uint64_t pn)
+{
+    quicly_sentmap_iter_t iter;
+    const quicly_sent_packet_t *sent;
+
+    quicly_loss_init_sentmap_iter(loss, &iter, now, quicly_spec_context.transport_params.max_ack_delay, 0);
+    while ((sent = quicly_sentmap_get(&iter))->packet_number != pn) {
+        assert(sent->packet_number != UINT64_MAX);
```

Comment:
Why isn't this `ASSERT`?

Reply:
Because this is not part of the loss recovery test. Rather, it is making sure that sentmap is behaving correctly (or that we are using the sentmap API correctly).